src/main/java/dev/galacticraft/mod/api/block/PipeShapedBlock.java

```diff
@@ -0,0 +1,180 @@
+
+    protected boolean updateConnection(BlockState currentState, BlockPos pos, Direction side, BlockPos neighborPos, Level level) {
+        if (level.getBlockEntity(pos) instanceof Connected pipe) {
+            boolean canConnect = this.canConnectTo(level, pos, side, neighborPos, currentState);
+
+            BlockState neighborState = level.getBlockState(neighborPos);
+            if (neighborState.getBlock() instanceof PipeShapedBlock<?> neighbor) {
+                canConnect &= neighbor.canConnectTo(level, neighborPos, side.getOpposite(), pos, neighborState);
+            }
+
+            boolean currentlyConnected = pipe.getConnections()[side.get3DDataValue()];
+            pipe.getConnections()[side.get3DDataValue()] = canConnect;
+            level.blockEntityChanged(pos);
+            return canConnect != currentlyConnected;
+        } else {
+            return false;
+        }
+    }
+
+    @Override
+    public void setPlacedBy(Level level, BlockPos pos, BlockState state, @Nullable LivingEntity placer, ItemStack itemStack) {
+        super.setPlacedBy(level, pos, state, placer, itemStack);
+
+        for (Direction direction : Direction.values()) {
+            this.updateConnection(state, pos, direction, pos.relative(direction), level);
+        }
+    }
+
+    @Override
+    public void neighborChanged(BlockState state, Level level, BlockPos pos, Block block, BlockPos neighborPos, boolean notify) {
+        super.neighborChanged(state, level, pos, block, neighborPos, notify);
+
+        Direction direction = Direction.fromDelta(neighborPos.getX() - pos.getX(), neighborPos.getY() - pos.getY(), neighborPos.getZ() - pos.getZ());
+        if (direction == null)
+            return;
+
+        if (this.updateConnection(state, pos, direction, neighborPos, level)) {
+            level.updateNeighborsAtExceptFromFacing(pos, state.getBlock(), direction);
+            this.onConnectionChanged(level, pos, direction, neighborPos);
+        }
+    }
+
+    @Override
+    protected @NotNull BlockState updateShape(BlockState state, Direction direction, BlockState neighborState, LevelAccessor levelAccessor, BlockPos pos, BlockPos neighborPos) {
+        if (levelAccessor instanceof Level level) {
+            if (this.updateConnection(state, pos, direction, neighborPos, level)) {
+                level.updateNeighborsAtExceptFromFacing(pos, state.getBlock(), direction);
+                this.onConnectionChanged(level, pos, direction, neighborPos);
+            }
+        }
+
+        return state;
+    }
+
+    @Override
+    protected @NotNull VoxelShape getShape(BlockState state, BlockGetter world, BlockPos pos, CollisionContext context) {
+        if (world.getBlockEntity(pos) instanceof Connected connected) {
+            return this.shapes[generateAABBIndex(connected)];
+        }
+        return this.shapes[0];
+    }
+
+    public static VoxelShape[] makeShapes(float radius) {
```

Comment:
I think this needs some special-casing for full-block pipes (e.g. sealable pipes/wires). Currently the bounds just explode out of the block.

Reply:
My plan is to have the full-block ones as their own thing separate from the PipeShapedBlock set of classes, since the block entity is really the only thing they have in common with the thin pipe-likes.